### PR TITLE
ARROW-7084: [C++] Check for full type equality in ArrayRangeEquals

### PR DESCRIPTION
**Pull request by tianchen92**
ArrayRangeEquals would only checks type IDs before comparing values. Make it check the actual type instances instead.

Also, fix the MapArray comparison logic to avoid accounting for field names.

### CODE REVIEW

File: cpp/src/arrow/compare.cc [hunk 1]
```diff
@@ -984,7 +984,8 @@ bool ArrayRangeEquals(const Array& left, const Array& right, int64_t left_start_
   bool are_equal;
   if (&left == &right) {
     are_equal = true;
-  } else if (left.type_id() != right.type_id()) {
+  } else if (left.type_id() != right.type_id() ||
+             !TypeEquals(*left.type(), *right.type(), false /* check_metadata */)) {
```
Comment (wesm) [inline cpp/src/arrow/compare.cc:988]: Can you add a couple of tests with some parametric types (e.g. TIMESTAMP with different time units)?

Comment (tianchen92) [reply cpp/src/arrow/compare.cc:988]: Thanks wes, fixed.

**Comment by tianchen92** [comment]
With this change, generate_non_canonical_map_case would fail, i think it is because the map field names. What is the right fix here? @pitrou

**Comment by pitrou** [comment]
@tianchen92 I'll take a look.

**Pull request closed without merging**